### PR TITLE
[bugfix] Fixes relation widget reference when filter value contains a quote #16399

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

This PR fixes the relation widget reference when filter value contains a quote (see https://issues.qgis.org/issues/16399).

Moreover, a unit test has been added to highlight the issue and there are other fixes about filtering safety.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/qgsrelation.cpp [hunk 1]
```diff
@@ -170,22 +170,23 @@ QString QgsRelation::getRelatedFeaturesFilter( const QgsFeature &feature ) const
...
+    const QString quotedColRef = QgsExpression::quotedColumnRef( fieldPair.referencingField() ) ;
 
     QVariant val( feature.attribute( fieldPair.referencedField() ) );
 
     if ( val.isNull() )
     {
-      conditions << QStringLiteral( "\"%1\" IS NULL" ).arg( fieldPair.referencingField() );
+      conditions << QStringLiteral( "%1 IS NULL" ).arg( quotedColRef );
     }
     else if ( referencingField.type() == QVariant::String )
```
Comment (nyalldawson) [inline src/core/qgsrelation.cpp]: I'd suggest getting rid of the separate handling for string/non string and always using quotedValue with the variant input. Then it'll automatically handle this extra logic, together with handling of Boolean/lists/other stuff in future.

Comment (pblottiere) [inline src/core/qgsrelation.cpp]: Done.

**Comment by pblottiere** [comment]
@nyalldawson Thank you for your review! 

I updated the code according to your comment.

**Comment by pblottiere** [comment]
I added a method `createFieldEqualityExpression` (not convinced by the name neither...). 

Let me know if it corresponds to what you had in mind.

**Comment by nyalldawson** [comment]
Looks good, and seems it'll already fix a few bugs. Can you add some tests for it too?

File: src/gui/editorwidgets/qgsrelationreferencesearchwidgetwrapper.cpp [hunk 2]
```diff
@@ -145,23 +145,8 @@ void QgsRelationReferenceSearchWidgetWrapper::onValueChanged( const QVariant &va
 
 void QgsRelationReferenceSearchWidgetWrapper::setExpression( QString exp )
 {
-  QgsSettings settings;
-  QString nullValue = QgsApplication::nullRepresentation();
   QString fieldName = layer()->fields().at( mFieldIdx ).name();
-
-  QString str;
-  if ( exp == nullValue )
```
Comment (nyalldawson) [inline src/gui/editorwidgets/qgsrelationreferencesearchwidgetwrapper.cpp]: Actually this logic looks a bit different - seems to allow entry of "null" string in the search widget wrapper to search for null values. Better not change this one.

**Comment by pblottiere** [comment]
@nyalldawson I added some tests on the new method `createFieldEqualityExpression`.

**Comment by nyalldawson** [comment]
Manually pushed, with followups at https://github.com/qgis/QGIS/commit/963e25e1ee1e1ac5aeff78d95545b6399de62063 and https://github.com/qgis/QGIS/commit/29c4fcfc16fd1d74b1d660efc14c137cc203fdcb (be careful to include these if backporting)

**Comment by pblottiere** [comment]
Mmmm I had corrected these issues (what you've done in followups commits)... It seems I made a mistake during my rebase/squash...

Thank you for having taken care of this @nyalldawson !

**Comment by m-kuhn** [comment]
The issue is still open in the bug tracker.

Best to include `Fix #16399` somewhere (i.e. description not in between) to have the ticked closed automatically and have a proper audit trail

**Comment by pblottiere** [comment]
> Best to include Fix #16399 somewhere next time (i.e. description not in between) to have the ticked closed automatically and have a proper audit trail.

@m-kuhn Yes I noticed that. Sorry for the inconvenience, it won't happen again...

**Pull request merged**